Turns Auto-Accent into a by-character Preference, Desnowflakes Preferences.dm

KasparoVy opened the pull request:
Required SQL Change:
ALTER TABLE `characters` ADD `autohiss` TINYINT(1) NOT NULL AFTER `gear`;

In-game changing of auto-accent level is no different.
This preference can be different on each of your characters.

As requested in https://nanotrasen.se/forum/topic/9880-auto-accent-preference-by-character/

:cl:
rscadd: You can now set your character's auto-accent at character creation. The verb still exists in the OOC tab.
/:cl:

As requested, this also desnowflakes preferences.dm by turning as many as possible of the species name checks into species datum flag checks to improve maintainability.

Primary use case of the above would be the following: Lets say we add tail markings for Tajara. We'll only have to give them the appropriate flag in `code/modules/mob/living/carbon/human/species/station.dm` on top of adding the sprites/definition for the markings because `code/modules/client/preference/preferences.dm` now checks the species' flags and not the species' name.

**_Big thanks to tigercat2000 for the suggestion of the preferences.dm desnowflaking and proofreading_**

KasparoVy commented:
Forgot to change the schema, editing that RN

tigercat2000 commented on code/modules/client/preference/preferences.dm:
```diff
@@ -289,6 +290,9 @@ var/global/list/special_role_times = list( //minimum age (in days) for accounts
 			if(species == "Grey")
 				dat += "<b>Voice:</b> <a href ='?_src_=prefs;preference=speciesprefs;task=input'>[speciesprefs ? "Wingdings" : "Normal"]</a><BR>"
 			dat += "<b>Secondary Language:</b> <a href='?_src_=prefs;preference=language;task=input'>[language]</a><br>"
+			if(species in list("Unathi", "Tajaran", "Plasmaman", "Kidan"))
```
please don't add more of these awful "species in list" checks, they make everything harder to add new stuff to. Add a `var/datum/species/S = all_species[species]` to the top of here and check something like `if(istype(S) && S.autohiss_basic_map)`.

KasparoVy replied:
Thanks for the tip

tigercat2000 commented on code/modules/mob/living/autohiss.dm:
```diff
@@ -10,7 +10,7 @@
-	if(!client || client.autohiss_mode == AUTOHISS_OFF) // no need to process if there's no client or they have autohiss off
+	if(!client || client.prefs.autohiss_mode == AUTOHISS_OFF) // no need to process if there's no client or they have autohiss off
 		return message
 	return species.handle_autohiss(message, L, client.autohiss_mode)
```
You didn't get rid of client.autohiss_mode, so the preference doesn't even work correctly

KasparoVy replied:
Works fine, just forgot to remove the original var after it reinstated in client preferences. Thanks for pointing it out

tigercat2000 commented on code/modules/client/preference/preferences.dm:
```diff
@@ -1366,6 +1373,12 @@ var/global/list/special_role_times = list( //minimum age (in days) for accounts
 
+				if("autohiss_mode")
+					if(species in list("Unathi", "Tajaran", "Plasmaman", "Kidan"))
+						var/list/autohiss_choice = list("Off" = 0, "Basic" = 1, "Full" = 2)
```
You're hardcoding the values of the defines here

KasparoVy replied:
Moving the autohiss define to the languages file to handle this

tigercat2000 replied:
You haven't replaced the hardcoded numbers with the defines here.

Aurorablade commented:
I don't agree with this unless I can be assured it won't cause a maintinability issue
 It's really better as a verb.

KasparoVy commented:
Just testing. I've addressed all of tiger's criticisms and I'm just making sure none of my changes broke anything.

I've changed as many species checks in preferences.dm as possible to be flag checks as requested, thus no maintainability issue.

Furthermore, this doesn't remove the verb. The verb is still there. This just allows you to set-and-forget at character creation so you don't have to remember to manually toggle it on at the beginning of every shift.

fludd12 commented:
You could do what I did with the grey wingding option, and make it use `speciesprefs`, since currently no race with autohiss uses that for anything. It would address @Aurorablade's concerns, and keep the hectic SQL testing to a minimum. Of course, this is entirely up to you, since having it as its own preference would allow for more varied autohissing.

KasparoVy commented:
The SQL is literally just a bolt-on. I avoided using `speciesprefs` in the event that gets used for something else (like was intended for a now-defunct PR). If it's what's desired over how this currently works, I'll change it. Otherwise, I see no problem.

KasparoVy commented:
@-spam me if it's something important. No idea why this is sitting and no-one's talked to me since I addressed the initial criticisms-- so don't be surprised if I don't respond immediately.

fludd12 commented:
It's taking a while to merge pulls, from what I can tell - been about a week since I had to fix any changes for the Grey rework.

KasparoVy commented:
No problem, was just curious
Thanks for helping to clear that up

fludd12 commented:
Not speaking officially, of course, just going by what I've seen
seriously take my word about github stuff with a grain of salt

Fox-McCloud commented:
@tigercat2000

KasparoVy commented:
:rofl:

KasparoVy commented:
Conflicts with #7137, resolving now.

KasparoVy commented:
Conflicts resolved

KasparoVy commented:
:thinking: Sounds good to me. I'll get that done in an hour or so.

KasparoVy commented:
There we are. Handy `random_character()` proc to the rescue.

tigercat2000 commented on code/modules/client/preference/preferences.dm:
```diff
@@ -1278,11 +1283,14 @@ var/global/list/special_role_times = list( //minimum age (in days) for accounts
 					species = input("Please select a species", "Character Generation", null) in new_species
 					var/datum/species/NS = all_species[species]
+					if(!istype(NS)) //The species was invalid. Set the species to the default and fetch the datum for that species.
+						species = initial(species)
```
In this case, I believe it would be better to reject the user's input, and inform them that whatever they chose is not a species.

tigercat2000 commented on code/modules/client/preference/preferences.dm:
```diff
@@ -1366,6 +1384,12 @@ var/global/list/special_role_times = list( //minimum age (in days) for accounts
 
+				if("autohiss_mode")
+					if(S.autohiss_basic_map)
+						var/list/autohiss_choice = list("Off" = 0, "Basic" = 1, "Full" = 2)
```
These still don't use the defines

tigercat2000 commented on code/modules/client/preference/preferences.dm:
```diff
@@ -1342,6 +1357,9 @@ var/global/list/special_role_times = list( //minimum age (in days) for accounts
 						rlimb_data = list()
+
+						if(!(NS.autohiss_basic_map))
+							autohiss_mode = 0
```
Nor does this

tigercat2000 commented on code/modules/client/preference/preferences.dm:
```diff
@@ -289,8 +296,10 @@ var/global/list/special_role_times = list( //minimum age (in days) for accounts
 				dat += "<b>Voice:</b> <a href ='?_src_=prefs;preference=speciesprefs;task=input'>[speciesprefs ? "Wingdings" : "Normal"]</a><BR>"
 			dat += "<b>Secondary Language:</b> <a href='?_src_=prefs;preference=language;task=input'>[language]</a><br>"
+			if(S.autohiss_basic_map)
+				dat += "<b>Auto-accent:</b> <a href='?_src_=prefs;preference=autohiss_mode;task=input'>[autohiss_mode == 2 ? "Full" : (autohiss_mode == 1 ? "Basic" : "Off")]</a><br>"
```
More missing defines

tigercat2000 commented on code/modules/mob/living/autohiss.dm:
```diff
@@ -1,37 +1,29 @@
-#define AUTOHISS_BASIC 1
-#define AUTOHISS_FULL 2
-
 #define AUTOHISS_NUM 3
```
This should probably go with the others for consistency

Fox-McCloud commented:
@Crazylemon64 couldja make the SQL changes for this, please?

KasparoVy commented:
😞 💤

Crazylemon64 commented on code/modules/client/preference/preferences.dm:
```diff
@@ -1068,7 +1080,7 @@ var/global/list/special_role_times = list( //minimum age (in days) for accounts
 				var/dflag=text2num(href_list["disability"])
 				if(dflag >= 0)
-					if(!(dflag==DISABILITY_FLAG_FAT && species!=("Human" || "Tajaran" || "Grey")))
+					if(dflag==DISABILITY_FLAG_FAT && (S.flags & CAN_BE_FAT))
```
This prevents any disability that is not obesity from being set

KasparoVy replied:
Yikes! I'll get on that right now. Thanks for pointing this out.
